rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,574 @@
+that there is limited UX degradation when compared to the a client connecting
+directly to the hardware key.
+
+##### Agent stops running
+
+If the agent stops unexpectedly during a dependent client's operation, it may
+lead to an error for the dependent client. Rather than returning this error,
+the client will log the error as debug and try again by connecting directly to
+the hardware key.
+
+#### Hardware Key prompts
+
+The agent is [responsible for hardware key PIN and touch prompts](#pin-and-touch-prompts).
+As a result, the dependent client will hang while it is prompted and handled
+in the agent client. Teleport Connect will foreground these prompts, so it
+should be clear to the user how to complete the prompts and proceed with the
+dependent client.
+
+Future improvement: instead of leaving the prompt purely up to the agent client,
+the agent client could propagate the prompt to the dependent client requesting
+a signature through a bi-directional, streaming version of the `Sign` rpc. The
+dependent could then prompt for PIN or touch like normal, e.g. in the terminal.
+
+### Proto
+
+#### `HardwareKeyAgentService`
+
+```proto
+// HardwareKeyAgentService provides an agent service for hardware key (PIV) signatures.
+// This allows multiple Teleport clients to share a PIV connection rather than blocking
+// each other, due to the exclusive nature of PIV connections. This also enables shared
+// hardware key states, such as a custom PIN cache shared across Teleport clients.
+service HardwareKeyAgentService {
+  // Ping the agent service to check if it is active.
+  rpc Ping(PingRequest) {PingResponse}
+  // Sign produces a signature with the provided options for the specified hardware private key
+  //
+  // This rpc implements Go's crypto.Signer interface.
+  rpc Sign(SignRequest) returns (Signature) {}
+  // GetInfo gets info for the specified hardware private key, including the public key,
+  // attestation statement, and supported pin and touch policies.
+  rpc GetInfo(GetInfoRequest) returns (GetInfoResponse) {}
+}
+
+// PingRequest is a request to Ping.
+message PingRequest {}
+
+// PingResponse is a response to Ping.
+message PingResponse {}
+
+// SignRequest is a request to perform a signature with a specific hardware private key.
+message SignRequest {
+  // KeyRef references a specific hardware private key.
+  KeyRef key_ref = 1;
+  // PublicKeyDer is a public key encoded in PKIX, ASN.1 DER form.
+  // This is provided to verify that the hardware private key has not
+  // been swapped out against the public key expected by the client.
+  bytes public_key_der = 2;
+  // Digest is a hashed message to sign.
+  bytes digest = 3;
+  // Hash is the hash function used to prepare the digest.
+  Hash hash = 4;
+  // SaltLength specifies the length of the salt added to the digest before a signature.
+  // Only used, and required, for PSS RSA signatures.
+  uint32 salt_length = 5;
```

Comment:
in the RSA lib a salt length of 0 indicates a default salt length so you may need to make this optional (i'm commenting on these because i have also implemented a signer over gRPC for vnet and i just realized we do PSS signatures and they currently don't work)

Reply:
https://pkg.go.dev/crypto/rsa#pkg-constants

Reply:
oh and it can be -1

Reply:
We're going to precompute the salt on the client side rather than waiting for the server to do it - here's the piv-go implementation that we'll essentially copy to the client side - https://github.com/go-piv/piv-go/blob/0383b0aa884b2b642e9e3446ea01ba22ccadc83a/v2/third_party/rsa/pss.go#L146

Reply:
We could also go back to having salt_length be an `int32`, allowing `[-1, max_int32)` (or IIRC `[-1, hash.size]`), or using a oneof that allows enum or uint, but I think precomputing just makes things more clear.

Reply:
i guess either is fine, i missed that you were going to precompute the actual salt length, is that mentioned somewhere?

Reply:
Added a note